thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,439 @@
+
+This is extremely useful for sending typed messages across channels. For example, in Servo we use such
+[an enum for sending details about the progress of a fetch to the corresponding XHR object][servo-xhr-enum].
+[Another such enum][servo-constellation-enum] is used for communication between the constellation and the compositor/script.
+
+
+This gives us a great degree of type safety; I can send messages with different data within them,
+however I can only send messages that the other end will know how to handle since they must all be of the type of the message enum.
+
+In Go there's no obvious way to get this. The closest thing is the type called `interface {}`
+which is similar to `Box<Any>` in Rust or `Object` in Java. This is a pointer to any type,
+with the ability to match on its type. As a Rustacean I felt incredibly dirty using this, since I expected that
+there would be an additional vtable overhead &mdash; in Rust `Box<Any>` or `Box<Trait>` are generally avoided in favor of
+generics (for compile time matching) or enums (for runtime matching). Besides, `interface{}` can be fed any type,
+so I can always accidentally send a message of the wrong type through a channel and
+it'll end up crashing the other end at runtime since it hit a `default:` case or something.
+
+So, for example, in Rust I design a messaging system for drawing to a window like so:
+
+```rust
+enum MyMessage {
+    Quit,
+    DrawLine(Color, Point, Point),
+    DrawTriangle(Color, Color, Point, Point, Point),
+    SetBackground(Color)
+    // Enum variants can also have named fields
+    // though this is behind a feature gate right now
+    // Either way, you can get named fields by creating a struct
+    // for each variant to contain
+}
+
+// On the receiver end
+fn painting_loop(rx: Receiver<MyMessage>) {
+    loop {
+        match rx.recv().unwrap() {
+            Quit => {
+                // shut down the window
+            },
+            DrawLine(color, a, b) => {
+                // draw a line from point `a` to point `b`
+                // with color `color`
+            },
+            DrawTriangle(color, fill, a, b, c) => {
+                // draw a triangle between points `a`, `b`, `c`
+                // with line color `color` and fill `fill`
+            },
+            SetBackground(color) => {
+                // Set background to `color`
+            }
+            // No need for a default case here since we handled
+            // all possible inputs
+            // If we had missed a message type, this would not compile
+            // without a wildcard case
+        }
+    }
+}
+
+// On the sender end
+fn draw_thingy(tx: Sender<MyMessage) {
+    tx.send(DrawLine(Red, Point{x:0, y:0}, Point{x: 1, y: 1}));
+    tx.send(SetBackground(Blue))
+    // we can't send anything other than `MyMessage` variants    
+}
+
+fn main() {
+    // do some setup
+    let (rx, tx) = channel();
```

Comment:
Just reading up on Rust as I edit, should this be:

`let (tx, rx)`?

https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html